DIPs/1NNN-MHH.md

```diff
@@ -0,0 +1,242 @@
+    mixin(format!"alias mod = %s;"(m));
+    //Your implementation goes here
+    pragma(msg, getSymbolsByUDA!(mod, FinderUDA));
+}
+
+@FinderUDA!"Hello" 
+void wow()
+{
+	import std.stdio;
+    writeln("WOW!");
+}
+```
+In this case it prints "tuple(wow)".
+
+This pattern has obvious flaws: Unless told where to look recursively, it can only work with symbols at module scope, the attribute declaration/s must be on separate lines to be uniquely identified, and more subtly in a big file we now have to search through every declaration that the UDA *could* be attached to. The compiler already has this information, let's use it.
+
+### How the patterns enabled by this DIP are done in C++
+Thanks to the simplicity of textual preprocessing, C++ can enable this pattern using a macro. However, this is extremely error-prone so a common way of achieving everything mentioned so far in this article is to have a separate compilation step to collect and use information from the source code.
+
+For example, Epic Games' Unreal Engine uses the following syntax to interface between C++ and "Blueprints" 
+```C++
+class AGameActor : public AActor
+{
+  GENERATED_BODY()
+  public:
+  UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = Properties)
+  FString Name;
+  UFUNCTION(BlueprintCallable, Category = Properties)
+  FString ToString();
+};
+```
+The game then inserts hijacks the compile to read the (in our terms) user-defined attributes from the header file, however this not only hurts compilation times but is also fairly antithetical to the "D way" of doing things.
+## Description
+This DIP simply proposes adding to the language a new *SpecialKeyword* `__ATTRIBUTE__` to the grammar. 
+```diff
+SpecialKeyword:
+    __FILE__
+    __FILE_FULL_PATH__
+    __MODULE__
+    __LINE__
+    __FUNCTION__
+    __PRETTY_FUNCTION__
++   __ATTRIBUTE__
+```
+It will always resolve to a string array literal.
+```D
+void main(string[] args)
+{
+    const attr = __ATTRIBUTE__;
+    pragma(msg, typeof(attr), attr); //"const(string[]) and []"
+}
+```
+This literal shall be empty unless `__ATTRIBUTE__` is used as a default initializer, in which case it shall be resolved to either to be either empty or an array literal of the fully qualified names of all declarations a UDA is attached to, if and only if (subject to existing default initializer resolution) it is resolved to an expression within said *UserDefinedAttribute*
+
+### Some specific examples:
+A templated struct
+```D
+module testmodule;
+struct Test(string name, string[] attr = __ATTRIBUTE__)
+{
+    pragma(msg, name, " says: ", attr);
+    this(int l) {/*Do work*/}  
+}
+@Test!"name"(1) 
+int echo(int x)
+{
+    return x;
+}
+//"name says: ["testmodule.echo"]"
+```
+A simple function
+```D
+module testmodule;
+auto just(string[] at = __ATTRIBUTE__)
+{
+    return at;
+}
+
+@(just())
+int x, y, z;
+
+pragma(msg, pragma(msg, __traits(getAttributes, x)[0]));
+//"["testmodule.x", "testmodule.y", "testmodule.z"]"
```

Comment:
This example is needed earlier establishing that a single UDA can refer to multiple declarations simultaneously.